src/expression.c

```diff
@@ -118,7 +118,12 @@ Expression *getRightThis(Loc loc, Scope *sc, AggregateDeclaration *ad,
                     {
                         //printf("rewriting e1 to %s's this\n", f->toChars());
                         n++;
+#ifdef IN_GCC
+                        if (n > 1)
+                            e1 = new VarExp(loc, f->vthis);
+#else
                         e1 = new VarExp(loc, f->vthis);
+#endif
```

Comment:
I think the only blocker is closures.  In some cases the frontend does not correctly infer the parent 'this'  (I raised this in the ML but no one responded).

```
void foo()
{
    class Inner  {  }  // Inner's parent chain is type void* as it points to foo's closure.
}
```
vs.
```
Class Outer
{
    void foo()
    {
        class Inner  {  }  // Inner's parent chain is type Outer, but the backend still populates it with foo's closure.
    }
}
```